akka-stream/src/main/scala/akka/stream/impl/fusing/IteratorInterpreter.scala

```diff
@@ -3,64 +3,71 @@
  */
 package akka.stream.impl.fusing
 
-import akka.event.NoLogging
+import akka.event.{ Logging, NoLogging }
 import akka.stream._
+import akka.stream.impl.fusing.GraphInterpreter.{ GraphAssembly, DownstreamBoundaryStageLogic, UpstreamBoundaryStageLogic }
+import akka.stream.stage.AbstractStage.PushPullGraphStage
 import akka.stream.stage._
 
 /**
  * INTERNAL API
  */
 private[akka] object IteratorInterpreter {
-  final case class IteratorUpstream[T](input: Iterator[T]) extends PushPullStage[T, T] {
-    private var hasNext = input.hasNext
 
-    override def onPush(elem: T, ctx: Context[T]): SyncDirective =
-      throw new UnsupportedOperationException("IteratorUpstream operates as a source, it cannot be pushed")
+  final case class IteratorUpstream[T](input: Iterator[T]) extends UpstreamBoundaryStageLogic[T] {
+    val out: Outlet[T] = Outlet[T]("IteratorUpstream.out")
+    out.id = 0
 
-    override def onPull(ctx: Context[T]): SyncDirective = {
-      if (!hasNext) ctx.finish()
-      else {
-        val elem = input.next()
-        hasNext = input.hasNext
-        if (!hasNext) ctx.pushAndFinish(elem)
-        else ctx.push(elem)
-      }
+    private var hasNext = input.hasNext
 
-    }
+    setHandler(out, new OutHandler {
+      override def onPull(): Unit = {
+        if (!hasNext) completeStage()
+        else {
+          val elem = input.next()
+          hasNext = input.hasNext
```

Comment:
ah, this was ported from a PushPull stage that requires push to be the last statement